contracts/exchange/contracts/src/mixins/MExchangeRichErrorTypes.sol

```diff
@@ -0,0 +1,109 @@
+/*
+
+  Copyright 2018 ZeroEx Intl.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+*/
+
+pragma solidity ^0.5.5;
+
+import "@0x/contracts-utils/contracts/src/mixins/MRichErrorTypes.sol";
+import "@0x/contracts-exchange-libs/contracts/src/LibOrder.sol";
+
+
+contract MExchangeRichErrorTypes is
+    MRichErrorTypes
+{
+    enum FillErrorCodes {
+        INVALID_TAKER_AMOUNT,
+        TAKER_OVERPAY,
+        OVERFILL,
+        INVALID_FILL_PRICE
+    }
+
+    enum SignatureErrorCodes {
+        BAD_SIGNATURE,
+        INVALID_LENGTH,
+        UNSUPPORTED,
+        ILLEGAL,
+        INAPPROPRIATE_SIGNATURE_TYPE
+    }
+
+    enum AssetProxyDispatchErrorCodes {
+        INVALID_ASSET_DATA_LENGTH,
+        UNKNOWN_ASSET_PROXY
+    }
+
+    enum TransactionErrorCodes {
+        NO_REENTRANCY,
+        ALREADY_EXECUTED
+    }
+
+    bytes4 internal constant SIGNATURE_ERROR_SELECTOR =
```

Comment:
Oh, wow, I had no idea. That makes no sense to me. Definitely worth changing.

Reply:
@hysz does the new `constantOptimizer` optimizer flag change this? I haven't really looked into it, but I'm not sure what else it would do (this is already set to `true` in our `compiler.json`).

Reply:
Fixed in 8243a6f